frame/plasma/src/lib.rs

```diff
@@ -0,0 +1,543 @@
+    ragne: Range<Balance>,
+    block_number: BlockNumber,
+    state_object: Property<AccountId>,
+}
+
+#[derive(Encode, Decode, Clone, Default, RuntimeDebug, PartialEq, Eq)]
+pub struct Checkpoint<AccountId, Balance> {
+    subsrange: Range<Balance>,
+    state_update: Property<AccountId>,
+}
+
+#[derive(Encode, Decode, Clone, Default, RuntimeDebug, PartialEq, Eq)]
+pub struct Exit<AccountId, Range, BlockNumber, Balance, Hash> {
+    state_update: StateUpdate<AccountId, Range, BlockNumber>,
+    inclusion_proof: InclusionProof<AccountId, Balance, Hash>,
+}
+
+#[derive(Encode, Decode, Clone, Default, RuntimeDebug, PartialEq, Eq)]
+pub struct InclusionProof<AccountId, Balance, Hash> {
+    address_inclusion_proof: AddressInclusionProof<AccountId, Balance, Hash>,
+    interval_inclusion_proof: IntervalInclusionProof<Balance, Hash>,
+}
+
+#[derive(Encode, Decode, Clone, Default, RuntimeDebug, PartialEq, Eq)]
+pub struct IntervalInclusionProof<Balance, Hash> {
+    leaf_index: Balance,
+    leaf_position: Balance,
+    siblings: Vec<IntervalTreeNode<Balance, Hash>>,
+}
+
+#[derive(Encode, Decode, Clone, Default, RuntimeDebug, PartialEq, Eq)]
+pub struct AddressInclusionProof<AccountId, Balance, Hash> {
+    leaf_index: AccountId,
+    leaf_position: Balance,
+    siblings: Vec<AddressTreeNode<AccountId, Hash>>,
+}
+
+#[derive(Encode, Decode, Clone, Default, RuntimeDebug, PartialEq, Eq)]
+pub struct IntervalTreeNode<Balance, Hash> {
+    data: Hash,
+    start: Balance,
+}
+
+#[derive(Encode, Decode, Clone, Default, RuntimeDebug, PartialEq, Eq)]
+pub struct AddressTreeNode<AccountId, Hash> {
+    data: Hash,
+    token_address: AccountId,
+}
+
+pub type BalanceOf<T> =
+    <<T as Trait>::Currency as Currency<<T as frame_system::Trait>::AccountId>>::Balance;
+pub type CheckpointOf<T> = Checkpoint<<T as frame_system::Trait>::AccountId, BalanceOf<T>>;
+pub type RangeOf<T> = Range<BalanceOf<T>>;
+pub type InclusionProofOf<T> = InclusionProof<
+    <T as frame_system::Trait>::AccountId,
+    BalanceOf<T>,
+    <T as frame_system::Trait>::Hash,
+>;
+pub type IntervalInclusionProofOf<T> =
+    IntervalInclusionProof<BalanceOf<T>, <T as frame_system::Trait>::Hash>;
+pub type IntervalTreeNodeOf<T> = IntervalTreeNode<BalanceOf<T>, <T as frame_system::Trait>::Hash>;
+pub type AddressInclusionProofOf<T> = AddressInclusionProof<
+    <T as frame_system::Trait>::AccountId,
+    BalanceOf<T>,
+    <T as frame_system::Trait>::Hash,
+>;
+pub type AddressTreeNodeOf<T> =
+    AddressTreeNode<<T as frame_system::Trait>::AccountId, <T as frame_system::Trait>::Hash>;
+pub trait PlappsAddressFor<Hash, AccountId> {
+    fn plapps_address_for(hash: &Hash, origin: &AccountId) -> AccountId;
+}
+
+/// Simple plapps address determiner.
+///
+/// Address calculated from the code (of the constructor), input data to the constructor,
+/// and the account id that requested the account creation.
+///
+/// Formula: `blake2_256(blake2_256(code) + blake2_256(data) + origin)`
```

Comment:
It's mean that I can't deploy same code/params twice, right? What's about `nonce` here?

Reply:
@akru Added the description of Plapps hash.

Reply:
@satellitex I just want to take your attention that this formula don't permit to make multiple plapps instances with similar parameters. May be it dont needed.

Reply:
That's no problem. Thanks.